### PR TITLE
Add exclude tests option

### PR DESCRIPTION
**Pull request by RadWolfie**
I recently realize exclude tests will be very useful for emulators to skip them if they are not compatible or aren't working by design. Exclude tests become necessary to verify other test results instead of manually inputting many individual tests, just to remove one or some tests.

---
In `config.txt` file:
`tests-exclude=<export hex number>[,...]`

### CODE REVIEW

**Comment by RadWolfie** [comment]
Converted to draft pull request because #97 pull request is top priority. Once it is merged, this pull request can leave draft state.

File: src/main.c
Comment (PatrickvL) [inline src/main.c]: Typo : exclude > excluded

File: src/main.c
Comment (PatrickvL) [inline src/main.c:146]: Contrary to the code above, it appears as if no tests are actually excluded in this loop?

Nevermind, all good - the exclusion is in `run_test()`.

File: src/main.c
Comment (PatrickvL) [inline src/main.c]: Same typo

File: src/main.c
Comment (PatrickvL) [inline src/main.c:117]: Since this can become an exponential-time check, it might be better to implement with a bit set instead of a vector, so that this search loop can be changed into a much faster index-based bit test.

File: src/main.c
Comment (PatrickvL) [inline src/main.c]: Nit: this variable does not really contain a remainder size. It contains an exclusion count in this code block.

File: src/main.c
Comment (PatrickvL) [inline src/main.c:98]: Would it be possible to alternatively specify tests to exclude _by name_ instead of (or in addition to) by number?

File: src/main.c
Comment (PatrickvL) [inline src/main.c:126]: `kernel_thunk_table` is not actually a kernel thunk table, right? It is an array of kernel API test functions, so perhaps it's clearer to update its name accordingly.

**Review by PatrickvL — CHANGES_REQUESTED** [review]
Some remarks to ponder...

File: src/main.c
Comment (RadWolfie) [reply src/main.c:126]: Technically, you are correct. I think I did not rename the variable. And I may not know a better wording for it. Unless it's better to refer as... `kernel_api_tests` variable?

File: src/main.c
Comment (RadWolfie) [reply src/main.c:98]: That will be a feature request and will be able to support in a separate branch. Which I had reworked the global tests to remove user error mistakes. Plus no longer need to manually add hex value and name of api strings for each tests. So, like I said, your suggestion is a feature request.

Currently, it's following same way for tests input.

File: src/main.c
Comment (RadWolfie) [reply src/main.c]: Hmm my thought are exclude is present and excluded would be past tense. Except grammar check says excluded because of "be". I will fix it.

File: src/main.c
Comment (RadWolfie) [reply src/main.c]: It's best to keep the wording different since `tests_exclude.size` and `(tests_)exclude_size` are too identical to cause confusion. Though, `exclude_count` can work as a replacement.

File: src/main.c
Comment (RadWolfie) [reply src/main.c:117]: integer (4 bytes) can only hold up to 32 bits and long long (8 bytes) can hold up to 64 bits. We have 378 kernel apis which need 6 of long long values. Plus this file is in C language which doesn't add such flexibility to extend such support.

Meaning we would have to write a utility to able support your suggestion. Unless there's something I do not know about? I rather avoid using third-party library nor create our own library for this type of complex.

Comment (PatrickvL) [reply src/main.c:117]: Maybe https://en.cppreference.com/w/cpp/utility/bitset helps

EDIT: scratch that. You already wrote this is C, not C++.

But, a bitset could be created, laid on top of a vector. Or if wasting memory doesn't matter much, each vector index could be associated with a test (when the int data is used as a boolean value)

Comment (RadWolfie) [reply src/main.c:117]: Yes, if this file was using C++ language, then I could had used the standard library's bitset class. Except it is not the main objective for this pull request. ergo720 did made a ticket #66 for suggestion change custom library to std::vector. This main.c file could change to C++ language to take advantage of C++ standard library usage.

Plus we also need to be sure that nxdk's custom libcxx can support bitset.

Keep in mind that nxdk's custom libcxx/pbclib are behind from upstream.

File: src/main.c
Comment (PatrickvL) [reply src/main.c:126]: Is this rename going to be part of this pull request, or in a later one?

Comment (RadWolfie) [reply src/main.c:126]: Later one, you'll know which one eventually. 😉

**Comment by RadWolfie** [comment]
Please mark conversations resolved if they are properly fixed. As for other questions/suggestion require later pull request(s), please keep them open. 🙂

**Review by PatrickvL — APPROVED** [review]

File: src/main.c
Comment (RadWolfie) [reply src/main.c:126]: resolved in #99 pull request.

File: src/main.c
Comment (RadWolfie) [reply src/main.c:98]: Transfer into ticket #100

File: src/main.c
Comment (RadWolfie) [reply src/main.c:117]: Transfer to ticket #101

**Pull request merged**
